[RFC] Multiple users in an event

webmat opened the pull request:
I'm opening this RFC with a target of stage 2. This is the first PR for this RFC, but a lot of discussion has already happened before on the subject. Notably in #809 and #589 (see this RFC's references links to for more past discussions).

Opening directly at stage 2 simply means we'll need to ensure all stage 1 and 2 criteria are satisfied before moving forward.

In this RFC I try to document all expected uses of the `user` fields. This is not meant to be only a description of the new fields that will be introduced by #869.

Since this RFC is meant to document all uses of the `user` fields, there will also be discussion about potentially deprecating and removing user fields in places that don't have a clear purpose. Feedback is definitely welcome, if you disagree with these removals.

**TODO**

- [x] Identify the sponsor
- [x] Populate the "source data" section
  - [x] add some simple Linux IAM events
  - [x] **I need help** to find more complex IAM event samples (e.g. listing changes to the user)

[Preview of the RFC](https://github.com/webmat/ecs/blob/multiple-users-rfc/rfcs/text/0007-multiple-users.md)

webmat commented:
This PR is no longer marked as Draft :-) 

Ping @ebeahan @epixa @jamiehynds @leehinman @janniten @willemdh @neu5ron @vbohata @rw-access @P1llus 

I'll be adding simple Linux events in the "Source Data" section.

However I would really appreciate if someone has real events they could share from AD or other, that capture some of the complex events with many users on the event (e.g. "real" user + priv escalation + target user + changes to the user).

jamiehynds commented:
@webmat 

**Event ID 4732 in AD - 'administrator' has added 'bob' to the users group.** 

A member was added to a security-enabled local group.
Subject:
   Security ID:  WIN-R9H529RIO4Y\Administrator
   Account Name:  Administrator
   Account Domain:  WIN-R9H529RIO4Y
   Logon ID:  0x1fd23

Member:
   Security ID:  WIN-R9H529RIO4Y\bob
   Account Name:  -

Group:
   Security ID:  BUILTIN\Users
   Group Name:  Users
   Group Domain:  Builtin

Additional Information:

   Privileges:  -

   Expiration time:  -

** Event ID 4722 in AD - 'administrator' has enabled the 'john.locke' account.** 

A user account was enabled.

Subject:

   Security ID:  ACME-FR\administrator
   Account Name:  administrator
   Account Domain:  ACME-FR
   Logon ID:  0x20f9d

Target Account:

   Security ID:  ACME-FR\John.Locke
   Account Name:  John.Locke
   Account Domain:  ACME-FR

janniten commented:
Hi All,
**AD Example**
I was not able to find an AD event where the 3 users appear. The relationship between users exists but it is scattered in
several events (4648,4624 and 4787) and bound but the winlog.logon.id. For example:
![image](https://user-images.githubusercontent.com/33020901/89998023-48e9c200-dc8d-11ea-9f6f-4a822ffadfd5.png)

I'm working now with some events in which all users maybe appears (4670,4715) in one event. I'll let you know If I found the example that contains all the users.

**Oracle Auditory Event Example**
 Those events are in the table dba_audit_trail in the SYS schema  (here the detailed structure: https://docs.oracle.com/database/121/REFRN/GUID-A9993FAC-12D3-4725-A37D-938CC32D74CC.htm#REFRN23023).
I prepare some small example, copying here the relevant columns but the full export is attached.

OS_USERNAME | USERNAME | ACTION_NAME | SQL_TEXT | Mapping
-- | -- | -- | -- | --
gabriel.surname | SYS | ALTER USER | ALTER USER ACRISTALDI   IDENTIFIED BY * | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI
gabriel.surname | SYS | SYSTEM GRANT | GRANT SELECT ANY   TABLE TO ACRISTALDI | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI
gabriel.surname | SYS | UPDATE | UPDATE SYS.USER$ SET   NAME='ACRISTAL' AND USER#=143 AND NAME='ACRISTALDI'; | **user.name:**   gabriel.surname     **user.effective:** SYS     **user.target:** ACRISTALDI     **user.changes:** ACRISTAL

[example_oracle.xlsx](https://github.com/elastic/ecs/files/5061746/example_oracle.xlsx)
 
**Privilege Access Management Logs**
In PAM solutions (with the exception of logs regarding to administration of secrets) we always have a case of privilege escalation,  so user.name and user.efective always exists and if the action is related to user management, then user.target and user.changes appears 
For example, I'm coping some headers from Cyberark's logs (unfortunatly I don't have access to export the logs)
![image](https://user-images.githubusercontent.com/33020901/90008932-72aae500-dc9d-11ea-97de-365b8ac39ca7.png)

I have access to some Thycotik logs, but I'm not able to generate logs. I'll be looking for an example there also.
Regards

webmat commented:
Amazing, thank you @janniten

leehinman commented:
AWS has an AssumedRole API, a simplified Cloudtrail entry would be:

``` json
{
  "eventName": "AssumeRole",
  "requestParameters": {
    "roleArn": "arn:aws:iam::111111111111:role/JohnRole2",
  },
  "resources": [
    {
      "ARN": "arn:aws:iam::111122223333:role/JohnRole2",
      "accountId": "111111111111",
      "type": "AWS::IAM::Role"
    }
  ],
  "responseElements": {
    "assumedRoleUser": {
      "arn": "arn:aws:sts::111111111111:assumed-role/test-role/Role2WithTags",
      "assumedRoleId": "AROAIFR7WHDTSOYQYHFUE:Role2WithTags"
    },
  "userIdentity": {
    "accessKeyId": "AKIAI44QH8DHBEXAMPLE",
    "accountId": "111111111111",
    "arn": "arn:aws:sts::111111111111:assumed-role/JohnDoe/JohnRole1",
    "principalId": "AROAIN5ATK5U7KEXAMPLE:JohnRole1",
    "type": "AssumedRole"
  }
}
```
| field             | value                               |
|-------------------|-------------------------------------|
| user.id           | AROAIN5ATK5U7KEXAMPLE:JohnRole1     |
| user.effective.id | AROAIFR7WHDTSOYQYHFUE:Role2WithTags |


Also AWS can have AssumedRoles in the userIdentity:

``` json
  "userIdentity": {
    "accessKeyId": "AKIAQRSTUVWXYZEXAMPLE",
    "accountId": "777788889999",
    "arn": "arn:aws:sts::777788889999:assumed-role/AssumeNothing/devdsk",
    "principalId": "AIDAQRSTUVWXYZEXAMPLE:devdsk",
    "sessionContext": {
      "attributes": {
        "creationDate": "2016-11-14T17:25:26Z",
        "mfaAuthenticated": "false"
      },
      "sessionIssuer": {
        "accountId": "777788889999",
        "arn": "arn:aws:iam::777788889999:role/AssumeNothing",
        "principalId": "AIDAQRSTUVWXYZEXAMPLE",
        "type": "Role",
        "userName": "AssumeNothing"
      }
    },
    "type": "AssumedRole"
  }
```
| field             | value                               |
|-------------------|-------------------------------------|
| user.id           | AIDAQRSTUVWXYZEXAMPLE               |
| user.name         | AssumeNothing                       |
| user.effective.id | AIDAQRSTUVWXYZEXAMPLE:devdsk        |

jamiehynds commented:
Event ID 4624 in Windows is used to capture when a process attempts to log on an account by explicitly specifying that account’s credentials.  This most commonly occurs in batch-type configurations such as scheduled tasks, or when using the RUNAS command.

A logon was attempted using explicit credentials.

Subject:
   Security ID:  WIN-R9H529RIO4Y\Administrator
   Account Name:  **Administrator**
   Account Domain:  WIN-R9H529RIO4Y
   Logon ID:  0x1ba0e
   Logon GUID:  {00000000-0000-0000-0000-000000000000}
Account Whose Credentials Were Used:
   Account Name:  **rsmith@mtg.com**
   Account Domain:  WIN-R9H529RIO4Y
   Logon GUID:  {00000000-0000-0000-0000-000000000000}
Target Server:
   Target Server Name: sp01.IceMAIL.com
   Additional Information: sp01.IceMAIL.com
Process Information:
   Process ID:  0x77c
   Process Name:  C:\Program Files\Internet Explorer\iexplore.exe
Network Information:
   Network Address: -
   Port:   -

webmat commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -0,0 +1,413 @@
+
+### Documenting the purpose of each usage of the user fields
+
+As of ECS 1.5, the ECS documentation doesn't have a good place to explain at length
```
Makes perfect sense, yes. I'll add that link.

jonathan-buttner commented:
@webmat just wanted to check in and see if you need me to solicit more feedback (I think I'm the sponsor for this one)?

webmat commented:
@jonathan-buttner Thanks for asking. Yes, a thorough review from someone on the team would be welcome.

Note that I still have a few tasks on my end before this is final, the main one being including concrete mapping examples based on Linux and Windows events (e.g. like those shared via comments here). This is unlikely to happen this week or the next.

So a review in the meantime would be greatly appreciated, with the only caveat that the concrete examples are coming :-)

jonathan-buttner commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -0,0 +1,419 @@
+#### Resolution
+
+Adding a way to document field sets via free form text is being worked on independently
+of this proposal ([ecs#943](https://github.com/elastic/ecs/issues/943)).
```
I wonder if it'd also be useful to have some `description` field or maybe `purpose` within the reusable section to document why the reuse is happening:

```YAML
# schemas/user.yml excerpt
  reusable:
    top_level: true
    expected:
      - client
      - destination
      - host
      - server
      - source
      # Added for this RFC
      - at: user
        as: target
        # field for documenting why the reuse is happening
        description: Whenever a user is performing an action that affects another user
          -- typically in IAM scenarios -- the user affected by the action is captured at user.target
      - at: user
        as: effective
      - at: user
        as: changes
```

webmat replied:
Yes we will handle that a bit more loosely. We have work in progress (https://github.com/elastic/ecs/issues/932) to support a section of free form asciidoc documentation per field set (in addition to all of the attributes in the yaml).

Once we have the mechanism for this free form documentation section, the "user" field set documentation will lay out all expected uses with examples, based on what's in this RFC.

jonathan-buttner replied:
Nice! Sounds good.

jonathan-buttner reviewed (dismissed):
The proposal looks good to me. @rw-access any additional scenarios you can think of that would warrant additional fields?

jonathan-buttner commented:
@rylnd not sure if you've seen anything from the SIEM that we'd need to include with these changes as well?

rylnd commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -0,0 +1,419 @@
+      "name": "bob.barker"
+    }
+  },
+  "related": { "user": ["alice", "root", "bob", "bob.barker"] }
```
Apologies if this has already been addressed, but one concern that was raised by the detections team was the uniqueness of these names. If `related.user` is intended to be used programmatically, my (naive) sense is that `id` would be a better candidate here.

webmat replied:
Yes, this has been raised in the past. We haven't addressed it directly anywhere yet, however. One potential solution that has been floated would be that `related.user` could perhaps contain all seen usernames + user IDs.

The field is only meant to pivot and find a user, no matter where in the event they appeared. It's not meant to have semantic meaning.

This RFC is not meant to address this issue, however. We will tackle that issue separately.

webmat commented:
Ok I think this is ready for a final review for stage 2.

@janniten Thanks so much for these events, they're the ones I used for the Windows examples. Note that it's fine if events don't actually need all the field nestings at the same time. They're all in the schema to support events that do need them all. But in many cases, the activity is spread out across multiple events, as you noted. So only the needed fields actually need to be used :-) I've tried to demonstrate the various situations in the examples I've just added. 

@leehinman Thanks a lot for the Cloudtrail AssumeRole events. I modified one of the ARNs slightly, but I used your events in my cloud example :-) Can you confirm my mapping makes sense to you? Especially the last paragraph.

Final note for reviewers: I took the liberty to use a "session" `event.category` in some of the examples. It doesn't exist yet in the ECS allowed values, but has been discussed in the [session RFC](https://github.com/elastic/ecs/pull/935). 

@ebeahan  We have one additional merge task once this is approved:

- [ ] Set the merge date
- [ ] Assign the RFC ID
  - [ ] *Rename the directory with field changes accordingly* :-)

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -282,6 +282,185 @@ the event now looks like:
+
+### Linux user creation
+
+Here's a typical set of log about a user creation on Linux.
```
```suggestion
Here's a typical set of logs about a user creation on Linux.
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -282,6 +282,185 @@ the event now looks like:
+{
+  "event": {
+    "category": ["process"],
+    "event": "creation"
```
```suggestion
    "type": "creation"
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -282,6 +282,185 @@ the event now looks like:
+{
+  "event": {
+    "category": ["session"],
+    "event": ["creation"],
```
```suggestion
    "type": ["creation"],
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -282,6 +282,185 @@ the event now looks like:
+{
+  "event": {
+    "category": ["session"],
+    "event": ["end"]
```
```suggestion
    "type": ["end"]
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -460,10 +463,106 @@ event 5:
+    "code": "4624",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "category": ["authentication"],
+    "event": ["start"],
```
```suggestion
    "type": ["start"],
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -460,10 +463,106 @@ event 5:
+    "code": "4781",
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "category": ["iam"],
+    "event": ["user", "change"],
```
```suggestion
    "type": ["user", "change"],
```

ebeahan commented on rfcs/text/0000-multiple-users.md:
```diff
@@ -571,7 +571,69 @@ Would translate to
+    "action": "AssumeRole",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
```suggestion
    "type": ["start"],
```

ebeahan commented on rfcs/0000-rfc-template.md:
```diff
@@ -83,7 +83,7 @@ The following are the people that consulted on the contents of this RFC.
 
 <!--
-Who will be or has consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.
+Who will be or has been consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.
```
👍

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+
+### Linux IAM and privilege escalation
+
+Here's a typical set of log about a user creation on Linux.
```
```suggestion
Here's a typical set of logs about a user creation on Linux.
```

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+  "event": {
+    "kind": "event",
+    "category": ["process"],
+    "event": "creation"
```
```suggestion
    "type": ["creation"]
```

webmat replied:
I will not drop a 🤦  on each of these, but the temptation is high 😂

Thanks for catching that.

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+  "event": {
+    "kind": "event",
+    "category": ["session"],
+    "event": ["creation"],
```
```suggestion
    "type": ["creation"],
```

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+  "event": {
+    "kind": "event",
+    "category": ["session"],
+    "event": ["end"]
```
```suggestion
    "type": ["end"]
```

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
```suggestion
    "type": ["start"],
```

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+    "provider": "Microsoft-Windows-Security-Auditing",
+    "kind": "event",
+    "category": ["iam"],
+    "event": ["user", "change"],
```
```suggestion
    "type": ["user", "change"],
```

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -0,0 +1,769 @@
+    "action": "AssumeRole",
+    "kind": "event",
+    "category": ["authentication"],
+    "event": ["start"],
```
```suggestion
    "type": ["start"],
```

ebeahan reviewed (approved):
LGTM 👍 

I only noted a handful of tiny changes.

ebeahan commented on rfcs/text/0007-multiple-users.md:
```diff
@@ -2,7 +2,7 @@
 
 - Stage: **2 (proposal)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
-- Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
+- Date: **2020-09-02** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
```
🎉

ebeahan reviewed (approved):
LGTM :shipit:

leehinman commented:
> Ok I think this is ready for a final review for stage 2.
> 
> @janniten Thanks so much for these events, they're the ones I used for the Windows examples. Note that it's fine if events don't actually need all the field nestings at the same time. They're all in the schema to support events that do need them all. But in many cases, the activity is spread out across multiple events, as you noted. So only the needed fields should be used in each such events :-) I've tried to demonstrate the various situations in the examples I've just added.
> 
> @leehinman Thanks a lot for the Cloudtrail AssumeRole events. I modified one of the ARNs slightly, but I used your events in my cloud example :-) Can you confirm my mapping makes sense to you? Especially the last paragraph.
> 
> Final note for reviewers: I took the liberty to use a "session" `event.category` in some of the examples. It doesn't exist yet in the ECS allowed values, but has been discussed in the [session RFC](https://github.com/elastic/ecs/pull/935).
> 
> @ebeahan We have one additional merge task once this is approved:
> 
> * [x]  Set the merge date
> * [x]  Assign the RFC ID
>   
>   * [x]  _Rename the directory with field changes accordingly_ :-)

LGTM.
